src/storage/object.ts

```diff
@@ -465,4 +470,42 @@ export class ObjectStorage {
       })
     )
   }
+
+  /**
+   * Generates a signed url for uploading an object
+   * @param objectName
+   * @param url
+   * @param expiresIn seconds
+   * @param owner
+   */
+  async signUploadObjectUrl(objectName: string, url: string, expiresIn: number, owner?: string) {
+    // check as super user if the object already exists
+    let found
+    try {
+      found = await this.findObject(objectName, 'id', true)
+    } catch (e) {
+      // Don't do anything, since it will throw an error if the object doesn't exist (but that is what we want.)
+    }
+
+    if (found) {
+      throw new StorageBackendError('Duplicate', 409, 'The resource already exists')
+    }
+
+    // check if user has INSERT permissions
+    await this.db.createObject({
```

Comment:
We would need to create a stored function that does something along these lines:


```
// canInsertObject

BEGIN

INSERT INTO objects ....

ROLLBACK
```

I have not tested the above, but the idea is to simply invoke an rpc function and run the insert to check the permissions and then roll it back if it's fine, else it will throw row-level error